basil/utils/utils.py

```diff
@@ -45,6 +46,61 @@ def bitarray_to_byte_array(bitarr):
     return array('B', bs.astype(np.uint8))
 
 
+def get_si_tcp(rel_path=''):
+    ''' Download SiTCP/SiTCP10G sources from the official github repo and apply patches
+    '''
+
+    sitcp_repo = r'https://github.com/BeeBeansTechnologies/SiTCP_Netlist_for_Kintex7'
+    sitcp_10G_repo = r'https://github.com/BeeBeansTechnologies/SiTCPXG_Netlist_for_Kintex7'
+
+    def line_prepender(filename, line):
+        with open(filename, 'rb+') as f:
+            content = f.read()
+            f.seek(0, 0)
+            add = bytearray()
+            add.extend(map(ord, line))
+            add.extend(map(ord, '\n'))
+            f.write(add + content)
+
+    sitcp_folder = os.path.join(os.path.os.getcwd(), rel_path, 'SiTCP/')
+
+    # Only download if the SiTCP git repository is not present
+    if not os.path.isdir(os.path.join(sitcp_folder, '.git')):
+        print('Downloading SiTCP')
+
+        # Has to be moved to be allowed to use existing folder for git checkout
+        git.Repo.clone_from(url=sitcp_repo,
+                            to_path=sitcp_folder, branch='master')
+        # Patch sources, see README of bdaq53
+        line_prepender(filename=sitcp_folder + 'TIMER.v', line=r'`default_nettype wire')
+        line_prepender(filename=sitcp_folder + 'WRAP_SiTCP_GMII_XC7K_32K.V', line=r'`default_nettype wire')
+        for line in fileinput.input([sitcp_folder + 'WRAP_SiTCP_GMII_XC7K_32K.V'], inplace=True):
+            print(line.replace("assign\tMY_IP_ADDR[31:0]\t= (~FORCE_DEFAULTn | (EXT_IP_ADDR[31:0]==32'd0) \t? DEFAULT_IP_ADDR[31:0]\t\t: EXT_IP_ADDR[31:0]\t\t);",
+                               'assign\tMY_IP_ADDR[31:0]\t= EXT_IP_ADDR[31:0];'), end='')
+    else:  # update if existing
+        print('SiTCP already present. Checking for updates')
+        g = git.cmd.Git(sitcp_folder)
+        g.pull()
+
+    sitcp_10G_folder = os.path.join(os.path.os.getcwd(), rel_path, 'SiTCP10G/')
```

Comment:
I would prefer a single function, but we can use a parameter string to specify which version/speed should be downloaded.

```python
def get_si_tcp(rel_path='', version={'1G','10G'}):
    ...
    if '1G' in version:
        ...
    if '10G' in version:
        ...
```

Reply:
I noticed that you check if the download already exists and is up to date. Since the download itself is small and quick, I am fine with keeping the one function as it is